File: postman/typescript/src/endpoints/apiEndpoints.ts
import { status } from "@gatling.io/http";
import { postman } from "@gatling.io/postman";

const collection = postman
  .fromResource("gatlingEcommerce.postman_collection.json")
  .environment("gatlingEcommerce.postman_environment.json");

const apiEndpoints = collection.folder("API Endpoints");

// Define session retrieval endpoint with response validation and data extraction
// Reference: https://docs.gatling.io/reference/script/core/checks/#validating
// Reference: https://docs.gatling.io/reference/script/core/checks/#extracting
export const session = apiEndpoints
  .folder("Authentication")
  .request("Create User Session")
  .check(status().is(200));

// Define products endpoint with pagination through query parameters and data extraction
export const products = apiEndpoints
  .folder("Products")
  .request("Get Products")
  .check(status().is(200));

// Define login request
// Reference: https://docs.gatling.io/reference/script/protocols/http/request/#forms
export const login = apiEndpoints
  .folder("Authentication")
  .request("User login")
  .check(status().is(200));

// Define search request
export const search = apiEndpoints
  .folder("Products")
  .request("Get Products")
  .check(status().is(200));

// Define the "Add to Cart" request with a JSON payload
// Reference: https://docs.gatling.io/reference/script/protocols/http/request/#elfilebody
export const cart = apiEndpoints.folder("Cart").request("Add to cart").check(status().is(200));

// Define checkout process
export const checkOut = apiEndpoints
  .folder("Checkout")
  .request("Checkout Order")
  .check(status().is(200));

File: kotlin/gradle/src/gatling/kotlin/example/endpoints/APIendpoints.kt
package example.endpoints

import example.utils.*
import io.gatling.javaapi.core.CoreDsl.*
import io.gatling.javaapi.http.*
import io.gatling.javaapi.http.HttpDsl.*
import java.util.Optional

// Add authentication header if an access token exists in the session
// Reference: https://docs.gatling.io/reference/script/protocols/http/protocol/#header
fun withAuthenticationHeader(protocolBuilder: HttpProtocolBuilder): HttpProtocolBuilder {
  return protocolBuilder.header("Authorization") { session ->
    Optional.ofNullable(session.getString(ACCESS_TOKEN)).orElse("")
  }
}

// Define session retrieval endpoint with response validation and data extraction
// Reference: https://docs.gatling.io/reference/script/core/checks/#validating
// Reference: https://docs.gatling.io/reference/script/core/checks/#extracting
val session: HttpRequestActionBuilder =
    http("Session")
        .get("/session")
        .check(status().`is`(200))
        .check(jmesPath("sessionId").saveAs(SESSION_ID))

// Define products endpoint with pagination through query parameters and data extraction
val products: HttpRequestActionBuilder =
    http("Product page: #{${PAGE_INDEX}}")
        .get("/products")
        .queryParam("page", "#{${PAGE_INDEX}}")
        .check(status().`is`(200))
        .check(jmesPath("products").saveAs(PRODUCTS))

// Define login request
// Reference: https://docs.gatling.io/reference/script/protocols/http/request/#forms
val login: HttpRequestActionBuilder =
    http("Login")
        .post("/login")
        .asFormUrlEncoded()
        .formParam("username", "#{username}")
        .formParam("password", "#{password}")
        .check(status().`is`(200))
        .check(jmesPath("accessToken").saveAs(ACCESS_TOKEN))

// Define search request
val search: HttpRequestActionBuilder =
    http("Search")
        .get("/products")
        .queryParam("search", "#{${PRODUCT_NAME}}")
        .check(status().`is`(200))

// Define the "Add to Cart" request with a JSON payload
// Reference: https://docs.gatling.io/reference/script/protocols/http/request/#elfilebody
val cart: HttpRequestActionBuilder =
    http("Add to Cart")
        .post("/cart")
        .asJson()
        .body(ElFileBody("bodies/cart.json"))
        .check(status().`is`(200))

// Define checkout process
val checkOut: HttpRequestActionBuilder =
    http("Checkout")
        .post("/checkout")
        .asJson()
        .body(ElFileBody("bodies/cart.json"))
        .check(status().`is`(200))
